sdk/core/azure-core/src/http/transport_policy.cpp

```diff
@@ -17,27 +17,104 @@ using namespace Azure::Core::Http;
 using namespace Azure::Core::Http::Policies;
 using namespace Azure::Core::Http::Policies::_internal;
 
-std::shared_ptr<HttpTransport> Azure::Core::Http::Policies::_detail::GetTransportAdapter()
-{
-  // The order of these checks is important so that WinHTTP is picked over libcurl on Windows, when
-  // both are defined.
+namespace Azure { namespace Core { namespace Http { namespace Policies { namespace _detail {
+  namespace {
+    bool AnyTransportOptionsSpecified(TransportOptions const& transportOptions)
+    {
+      return !(
+          transportOptions.HttpProxy.empty() && transportOptions.ProxyPassword.empty()
+          && transportOptions.ProxyUserName.empty());
+    }
+    //    std::once_flag createTransportOnce;
+    //    std::shared_ptr<HttpTransport> defaultTransport;
+  } // namespace
+
+  std::shared_ptr<HttpTransport> GetTransportAdapter(TransportOptions const& transportOptions)
+  {
+    // The order of these checks is important so that WinHTTP is picked over libcurl on
+    // Windows, when both are defined.
 #if defined(BUILD_TRANSPORT_CUSTOM_ADAPTER)
-  return ::AzureSdkGetCustomHttpTransport();
+    return ::AzureSdkGetCustomHttpTransport();
 #elif defined(BUILD_TRANSPORT_WINHTTP_ADAPTER)
-  return std::make_shared<Azure::Core::Http::WinHttpTransport>();
+    // Since C++11: If multiple threads attempt to initialize the same static local variable
+    // concurrently, the initialization occurs exactly once. We depend on this behavior to ensure
+    // that the singleton defaultTransport is correctly initialized.
+    static std::shared_ptr<HttpTransport> defaultTransport(std::make_shared<WinHttpTransport>());
+    if (AnyTransportOptionsSpecified(transportOptions))
+    {
+      WinHttpTransportOptions httpOptions;
+      if (!transportOptions.HttpProxy.empty())
+      {
+        std::string proxyString;
+        proxyString = "http=" + transportOptions.HttpProxy;
+        proxyString += ";";
+        proxyString += "https=" + transportOptions.HttpProxy;
+        httpOptions.ProxyInformation = proxyString;
+      }
+      httpOptions.ProxyUserName = transportOptions.ProxyUserName;
+      httpOptions.ProxyPassword = transportOptions.ProxyPassword;
+      return std::make_shared<Azure::Core::Http::WinHttpTransport>(httpOptions);
+    }
+    else
+    {
+      //      std::call_once(createTransportOnce, []() {
+      //      defaultTransport = std::make_shared<Azure::Core::Http::WinHttpTransport>();
+      //    });
+      return defaultTransport;
+    }
 #elif defined(BUILD_CURL_HTTP_TRANSPORT_ADAPTER)
-  return std::make_shared<Azure::Core::Http::CurlTransport>();
+    static std::shared_ptr<HttpTransport> defaultTransport(std::make_shared<CurlTransport>());
+    if (AnyTransportOptionsSpecified(transportOptions))
+    {
+      CurlTransportOptions curlOptions;
+      curlOptions.EnableCurlTracing = true;
+      if (!transportOptions.HttpProxy.empty())
+      {
+        curlOptions.Proxy = transportOptions.HttpProxy;
```

Comment:
We need to allow setting the proxy as an empty string for libcurl, as that's a why to ask libcurl to do-not-use any proxy.
For example, when proxy is set with environment variables, we need to allow per-client proxy override with an empty string.
We had an issue about this, not so long ago and I had to update the proxy options string to Nullable<T>

Now, if for winHttp an empty string doesn't mean to do-not-use a proxy, it means that the proxy option from the common transport adapter will have different meaning depending on the transport adapter... so it won't be very common after all.  Do you know what it means for winHttp to use empty string for proxy?

Reply:
An empty string for proxy for WinHTTP throws an error.

Proxies are disabled for WinHTTP, you need to explicitly enable them. The HTTP_PROXY/HTTPS_PROXY convention doesn't work for WinHTTP as far as I know.

Fixing to make the common options a `nullable<string>`